### PR TITLE
compute/persist_sink: efficient handling of future updates

### PR DESCRIPTION
**Pull request by teskje**
This PR makes the compute `persist_sink` more efficiently handle updates for future times by moving received updates into a stash first and then moving them into the `correction` buffer only when the input frontier has passed their times. The result is that we can ignore updates for future times in the process of determining which updates need to be sinked to persist. This has the potential for avoiding a large amount of unnecessary work, particularly in dataflows that contain temporal filters.

### Motivation

 * This PR fixes a recognized bug.

Fixes https://github.com/MaterializeInc/materialize/issues/25844

### Tips for reviewer

I think long-term we want to move to some more sophisticated merge batcher implementation that also enables us to spill to lgalloc, but this seems like a good short-term fix for the temporal filter issue.

See performance test results in the comments.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/A

### CODE REVIEW

**Comment by teskje** [comment]
To validate that this actually solves the issue and doesn't introduce new regressions, I ran two comparison tests. The first one uses [the temporal filter repro from #25844](https://github.com/MaterializeInc/materialize/issues/25844#issuecomment-1985688807), the second one uses TPC-H with pk/fk indexes and materialzied views on all of the queries.

In all graphs below, green is main, yellow is this branch.

# Temporal Filter Workload

The stash is effective in resolving the CPU usage issue. This has visible positive effects on responsiveness.

Screenshot 2024-03-11 at 11 45 27
Screenshot 2024-03-11 at 11 48 21

Steady-state physical memory usage is slightly increased. However, peak memory usage is the same.

Screenshot 2024-03-11 at 11 46 38
Screenshot 2024-03-11 at 11 50 41

File: src/compute/src/sink/persist_sink.rs
Comment (petrosagg) [inline src/compute/src/sink/persist_sink.rs:735]: this is probably clearer with a `match desired_stash.as_mut()` statement

**Review by petrosagg — APPROVED** [review]
Makes sense! I would personally not bother with a flag since this is clearly a beneficial change and I don't like to keep around conditional code ~forever. I'm not the maintainer of this code though so this is not a strong opinion

File: src/compute/src/sink/persist_sink.rs
Comment (frankmcsherry) [inline src/compute/src/sink/persist_sink.rs:738]: I think you may want to consolidate `stash.entry(t)` under some circumstances. Otherwise retractions of future emissions will linger and consume resources until that time arrives. Standard pattern is to consolidate on each 2x growth we experience, which you can often do by sneaking a peek at `Vec::capacity()`: consolidate when "full" and post-consolidation double the capacity if it is not at least half-empty.

File: src/compute/src/sink/persist_sink.rs
Comment (bkirwi) [inline src/compute/src/sink/persist_sink.rs:752]: It seems a bit wasteful to revisit all timestamps every time the frontier advances, even though:
- for the case that inspired this PR (IIUC) this may be quite a lot of distinct timestamps;
- the btreemap is stored in timestamp order, so in theory it should be efficient to peel off just the smallest timestamps.

I think if our timestamps were partially ordered we might just have to eat this cost; but I think here we're working with our concrete totally-ordered timestamp type and we can just pop off the entries in order. Does that scan?

**Review by bkirwi — APPROVED** [review]
Pretty compelling charts! Left one additional performance thought inline.

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [reply src/compute/src/sink/persist_sink.rs:738]: I was wondering about that! I wasn't sure if we expect to see many future insertions/retractions that cancel each other out in Materialize. But I guess there is not much harm in adding this optimization, apart from some additional code complexity.

**Comment by teskje** [comment]
> I would personally not bother with a flag since this is clearly a beneficial change and I don't like to keep around conditional code ~forever.

No plans to keep the flag around forever! But I do want to have it in case we run into regressions, however unlikely they might be. It's easy to add and easy to remove too.

I can be convinced to have the flag enabled by default though, if everyone thinks this change is not risky. Makes the rollout a bit simpler.

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [reply src/compute/src/sink/persist_sink.rs:752]: We indeed only sink totally-ordered times to persist. However, AFAICT, the whole `persist_sink` is currently written to accept partially-ordered times, and I'd rather not change that assumption as part of this PR, mostly because I think it might be controversial.

Definitely something to keep it mind as a potential follow-up though!

Comment (bkirwi) [reply src/compute/src/sink/persist_sink.rs:752]: Fair enough! (Given that all the operators in this file hardcode to `mz_repr::Timestamp`, it's unclear to me to what extent this is true in practice... but also I can imagine it being controversial anyways, and it's certainly not worth holding up a PR over. 😅)

Comment (teskje) [reply src/compute/src/sink/persist_sink.rs]: I added this optimization now after all, because you are right that we use `mz_repr::Timestamp` everywhere already and it was easy to add.

File: src/compute/src/sink/persist_sink.rs
Comment (antiguru) [inline src/compute/src/sink/persist_sink.rs:843]: We might not have enough information here, but if we had the input's frontier, we could skip inserting the data in the stash if we know that it's due to be `drain_before`d immediately.

File: src/compute/src/sink/persist_sink.rs
Comment (antiguru) [inline src/compute/src/sink/persist_sink.rs:860]: Uncertain if this is correct: We might be able to advance the times to the meet of persist and desired frontier to enable more consolidation immediately. I think we should leave it as-is for the moment but note it down as a path towards further optimization, unless this information just isn't available here.

File: src/compute/src/sink/persist_sink.rs
Comment (antiguru) [inline src/compute/src/sink/persist_sink.rs:628]: Comment please :) "Pushes `item` into the vector. If the vector does not have sufficient capacity, we try to consolidate and/or double the vector's capacity. The cost of this function is $O(n \log n)$ in the number of items the vector stores."

File: src/compute/src/sink/persist_sink.rs
Comment (antiguru) [inline src/compute/src/sink/persist_sink.rs:603]: You could call this `seal`, similarly to what the merge batcher offers. No strong opinion tho.

File: src/compute/src/sink/persist_sink.rs
Comment (antiguru) [inline src/compute/src/sink/persist_sink.rs:605]: Elsewhere we'd use an `AntichainRef`, but no need to change since this is just an internal method.

File: src/compute/src/sink/persist_sink.rs
Comment (antiguru) [inline src/compute/src/sink/persist_sink.rs:672]: This has the potential deficit that it costs `self.push` for every item, unless we can pre-size the inner vector. Mitigated by `push` being $O(1)$ amortized?

**Review by antiguru — APPROVED** [review]
I think this looks great, I left some comments inline. I think this PR has the side-effect that it increases the number of allocations and doesn't try to reuse any of them, but that seems like an OK trade off given the complexity it would introduce, i.e., essentially implementing a special merge batcher.

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [reply src/compute/src/sink/persist_sink.rs:843]: Yep, I don't think the async operator lets us get that information.

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [reply src/compute/src/sink/persist_sink.rs:672]: We can maybe do better by figuring out the sizes of batches that we push here. I didn't try to go there because (a) as you say, `push` will be O(1) usually and (b) `ConsolidateBuffer` does the same thing and seems fine with it. I think working with batches of known sizes makes the code more complicated (both the `ConsolidatingVec::push` and `UpdateStash::insert`) so I'd like to leave it as is unless we notice performance issues.

File: src/compute/src/sink/persist_sink.rs
Comment (teskje) [reply src/compute/src/sink/persist_sink.rs:605]: I'm leaving it like it is because it avoids one `.borrow()` at the call site, but something to reconsider if we ever want to use this type more widely.

**Pull request merged**
